pkg/controller/replicaset/replica_set_test.go

```diff
@@ -395,53 +331,6 @@ func TestStatusUpdatesWithoutReplicasChange(t *testing.T) {
 	fakeHandler.ValidateRequest(t, testapi.Extensions.ResourcePath(replicaSetResourceName(), rs.Namespace, rs.Name)+"/status", "PUT", &updatedRc)
 }
 
-func TestControllerUpdateReplicas(t *testing.T) {
-	// This is a happy server just to record the PUT request we expect for status.Replicas
-	fakeHandler := utiltesting.FakeHandler{
-		StatusCode:    200,
-		ResponseBody:  "{}",
-		SkipRequestFn: skipListerFunc,
-	}
-	testServer := httptest.NewServer(&fakeHandler)
-	defer testServer.Close()
-
-	client := clientset.NewForConfigOrDie(&restclient.Config{Host: testServer.URL, ContentConfig: restclient.ContentConfig{GroupVersion: &api.Registry.GroupOrDie(v1.GroupName).GroupVersion}})
-	stopCh := make(chan struct{})
-	defer close(stopCh)
-	manager, informers := testNewReplicaSetControllerFromClient(client, stopCh, BurstReplicas)
-
-	// Insufficient number of pods in the system, and Status.Replicas is wrong;
-	// Status.Replica should update to match number of pods in system, 1 new pod should be created.
-	labelMap := map[string]string{"foo": "bar"}
-	extraLabelMap := map[string]string{"foo": "bar", "extraKey": "extraValue"}
-	rs := newReplicaSet(5, labelMap)
-	rs.Spec.Template.Labels = extraLabelMap
-	informers.Extensions().V1beta1().ReplicaSets().Informer().GetIndexer().Add(rs)
-	rs.Status = extensions.ReplicaSetStatus{Replicas: 2, FullyLabeledReplicas: 6, ReadyReplicas: 2, AvailableReplicas: 2, ObservedGeneration: 0}
-	rs.Generation = 1
-	newPodList(informers.Core().V1().Pods().Informer().GetIndexer(), 2, v1.PodRunning, labelMap, rs, "pod")
-	newPodList(informers.Core().V1().Pods().Informer().GetIndexer(), 2, v1.PodRunning, extraLabelMap, rs, "podWithExtraLabel")
-
-	// This response body is just so we don't err out decoding the http response
-	response := runtime.EncodeOrDie(testapi.Extensions.Codec(), &extensions.ReplicaSet{})
-	fakeHandler.SetResponseBody(response)
-
-	fakePodControl := controller.FakePodControl{}
-	manager.podControl = &fakePodControl
-
-	manager.syncReplicaSet(getKey(rs, t))
-
-	// 1. Status.Replicas should go up from 2->4 even though we created 5-4=1 pod.
-	// 2. Status.FullyLabeledReplicas should equal to the number of pods that
```

Comment:
Do you have an integration test that checks FullyLabeledReplicas when it is not the same as Replicas?

Reply:
no, let me add it. thanks for catching it!

Reply:
@enisoc I managed to get all tests passed except this. In the test, I created a RS with selectors and labels as `map[extraKey:extraValue foo:bar]`. After the RS created its own pods, I changed one of the pod to have labels `map[foo:bar]`. 

I was able to verify the labels changed in the pooling loop:
```
before:  map[extraKey:extraValue foo:bar] map[extraKey:extraValue foo:bar]
after: map[foo:bar] map[extraKey:extraValue foo:bar]
after: map[foo:bar] map[extraKey:extraValue foo:bar]
...
```
However, `.Status.FullyLabeledReplicas` did not change (stayed at 2). Does the test algorithm looks correct to you?

Reply:
The test should look like this:

* Create RS with only labels {foo: bar} and 2 replicas. Wait for it to be stable.
* Change RS Pod template labels to {foo: bar, extra: extra}. But *don't* change the selector.
* Change the labels on only 1 Pod to {foo: bar, extra: extra}.
* Wait for RS Status to show Replicas=2, FullyLabeledReplicas=1.

Reply:
Thanks! I verify the algorithm is working and the test is passing. :)
I never thought about don't change the selector - the selector has to match template labels for apps/v1beta2, but then I realized the testing framework still uses extensions/v1beta1. I will add a note about the test may break if the testing framework migrates to use newer API version.

Reply:
The test shouldn't break for newer API versions. We require the selector to match the template in all versions.

When we say "match" we don't mean the set of labels has to be equal. We mean the selector will select objects with those labels.

For example, a selector of `{foo: bar}` "matches" an object with labels `{foo: bar, extra: extra}`.

Reply:
I got it - it is fine for selector to "match" some of the labels, not necessarily all.